storage,kvserver: fix bugs and tests when using separated intents

sumeerbhola opened the pull request:
- ClearIntent calls now pass the txn UUID.
- MVCC iteration is prevented from crossing into
  the lock table key space.
  pebbleIterator tracks the direction of iteration
  and cheaply checks whether it has crossed into
  the lock table key space, and if so, marks itself
  done.
  There is additional testing using intentInterleavingIter
  that exercises this code.
- The MVCCIterator interface is not properly designed
  for propagating errors, and it eats the error if the
  MVCCKey cannot be parsed. There is now a TODO to fix
  this.
- storage and kvserver test fixes.

Separated intents are not enabled in this PR. They were
temporarily enabled only to fix bugs and tests.

Informs #41720

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57507)
<!-- Reviewable:end -->

sumeerbhola committed:
storage,kvserver: fix bugs and tests when using separated intents

- ClearIntent calls now pass the txn UUID.
- MVCC iteration is prevented from crossing into
  the lock table key space.
  pebbleIterator tracks the direction of iteration
  and cheaply checks whether it has crossed into
  the lock table key space, and if so, marks itself
  done.
  There is additional testing using intentInterleavingIter
  that exercises this code.
- kvserver.checkCanInitializeEngine now uses an EngineIterator
  and the incorrect optimization to only iterate over
  10 keys is removed.
- The MVCCIterator interface is not properly designed
  for propagating errors, and it eats the error if the
  MVCCKey cannot be parsed. There is now a TODO to fix
  this.
- storage and kvserver test fixes.

Separated intents are not enabled in this PR. They were
temporarily enabled only to fix bugs and tests.

Informs #41720

Release note: None

nvb reviewed (commented):
> Reviewed 26 of 26 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57507)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/client_merge_test.go, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTRwvJAu6jwYUvCe2U:-MOTRwvJAu6jwYUvCe2V:be5qtxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/client_merge_test.go#L135)):*
> ```Go
> func getEngineKeySet(t *testing.T, e storage.Engine) map[string]struct{} {
> 	t.Helper()
> 	// Have to scan local and global keys separately.
> ```

Could you mention why this is? Maybe point at the code that enforces this limitation, in case the limitation is ever lifted.
___
*[pkg/kv/kvserver/store.go, line 2163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTS_4tA6StYlatgact:-MOTS_4tA6StYlatgacu:b-m2k8r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2163)):*
> ```Go
> // cluster version, which must be present.
> func checkCanInitializeEngine(ctx context.Context, eng storage.Engine) error {
> 	// We use an EngineIterator to ensure that there are no non-MVCC keys
> ```

Non-mvcc meaning lock-table keys? Does this include keys with a 0 timestamp?
___
*[pkg/kv/kvserver/store.go, line 2165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTTGZC5UlEwKxWCMuk:-MOTTGZC5UlEwKxWCMul:bbo91nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2165)):*
> ```Go
> 	// We use an EngineIterator to ensure that there are no non-MVCC keys
> 	// in the engine.
> 	const maxKeys = 10
> ```

What's the point of this 10 key limit? We just want to collect some keys, but not all, so we can fail if any of them are not the `StoreClusterVersionKey` or a `StoreCachedSettingsKey`?

The latter part (which was added in #55166) is a little suspect. Can we have more than 10 `StoreCachedSettingsKey`? If so, would that prevent us from noticing that there are other keys in the engine? cc. @tbg 
___
*[pkg/storage/intent_interleaving_iter.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTZHLd1Io3eTv623MN:-MOTZHLd1Io3eTv623MO:bmp72qt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/intent_interleaving_iter.go#L146)):*
> ```Go
> 	// TODO(sumeer): the creation of these iterators can race with concurrent
> 	// mutations, which may make them inconsistent with each other. Add a way to
> 	// clone the underlying Pebble iterator and use it in both places.
> ```

:thumbsup: cloning just means creating two iterators with the same sequence number, right?
___
*[pkg/storage/mvcc_incremental_iterator_test.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVA0C0RJabvzQjL_O:-MOTVA0C0RJabvzQjL_P:bphb9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/mvcc_incremental_iterator_test.go#L193)):*
> ```Go
> 
> 	var (
> 		keyMin   = keys.LocalMax
> ```

Should we rename `keyMin` to `localMin` to avoid the possibility of confusion?
___
*[pkg/storage/mvcc_test.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVlYzDQeyYyI_xs-H:-MOTVlYzDQeyYyI_xs-I:b-85csy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/mvcc_test.go#L48)):*
> ```Go
> // Constants for system-reserved keys in the KV map.
> var (
> 	keyMin       = keys.LocalMax
> ```

Same question here about `localMin`.
___
*[pkg/storage/mvcc_test.go, line 1037 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVvWc0QECtK9kTqh0:-MOTVvWc0QECtK9kTqh1:b-iy36tc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e926b4b04da57171d0fb38159a602da3a2f19f/pkg/storage/mvcc_test.go#L1037)):*
> ```Go
> 				t.Fatal(err)
> 			}
> 			// Should Next() from key2/5 to key2/3 first, then Seek to key3, and see
> ```

Why did this need to change? Something to do with the MVCCDelete and its interaction with the iterator?
___
*[pkg/storage/pebble_iterator.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTWIPK5lXVBs3eLuVB:-MOTWIPK5lXVBs3eLuVC:b-sz0lmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/pebble_iterator.go#L59)):*
> ```Go
> 	//
> 	// The current direction. +1 for forward, -1 for reverse.
> 	mvccDir int
> ```

nit: Is there a benefit to the int vs. a bool? I ask because the int is going to increase the size of this struct while a bool would pack into the same word as `mvccDone`. In fact, you could rearrange this struct to keep all the bools next to each other if you wanted to try to keep the total size down. 
___
*[pkg/storage/pebble_iterator.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTXjJg46yPgW1JUgad:-MOTXjJg46yPgW1JUgae:bedd55z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/pebble_iterator.go#L246)):*
> ```Go
> 		// not in the lock table key space.
> 		//
> 		// TODO(sumeer): fix this properly by changing those method signatures.
> ```

What is needed to fix the signatures? An error return value on `Key` and `UnsafeKey`? Or do you plan to pull key parsing into `Valid` for cases where it's needed?
___
*[pkg/storage/testdata/intent_interleaving_iter/basic, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTYjv2EzpfipUqZzUo:-MOTYjv2EzpfipUqZzUp:b-pn2qbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/testdata/intent_interleaving_iter/basic#L448)):*
> ```
> 
> # Confirm that the MVCCIterator does not iterate into the lock table key space
> # despite no upper bound.
> ```

Is this comment correct? Isn't this confirming that the _lock table iterator_ does not iterate _out of_ the lock table key space despite no upper bound?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57507)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @nvanbenschoten, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/store.go, line 2165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTTGZC5UlEwKxWCMuk:-MO_n6fWClv9dgk2qznY:b-hdgske) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2165)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's the point of this 10 key limit? We just want to collect some keys, but not all, so we can fail if any of them are not the `StoreClusterVersionKey` or a `StoreCachedSettingsKey`?

The latter part (which was added in #55166) is a little suspect. Can we have more than 10 `StoreCachedSettingsKey`? If so, would that prevent us from noticing that there are other keys in the engine? cc. @tbg 
</blockquote></details>

Yes, there can be more than 10. These keys mirror `system.settings`.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-:-MOc1Mfu8js2kWjZTJxB:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @nvanbenschoten, @petermattis, @sumeerbhola, and @tbg)

___
*[pkg/kv/kvserver/client_merge_test.go, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTRwvJAu6jwYUvCe2U:-MObaGa_5q9joRTv3lqZ:b-m25nad) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/client_merge_test.go#L135)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you mention why this is? Maybe point at the code that enforces this limitation, in case the limitation is ever lifted.
</blockquote></details>

I've added a comment here and in the declaration of `engine.Scan`.
___
*[pkg/kv/kvserver/consistency_queue_test.go, line 446 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MObc2K86qt6da1U-2qz:-MObc2K86qt6da1U-2r-:b9u4pmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/consistency_queue_test.go#L446)):*
> ```Go
> 		defer iter.Close()
> 
> 		ms, err := storage.ComputeStatsForRange(iter, keys.LocalMax, roachpb.KeyMax, 0 /* nowNanos */)
> ```

I've added a TODO/question in `mvcc.go` where `ComputeStatsForRange` is declared. If we consider the first KV range to start at `KeyMin`, then this change in tests may not be sufficient and we may have an error-prone situation whenever we add new callers for any function that expects the range to be only global or only local.

For functions like `ComputeStatsForRange` it is viable to do this compensation at the callee since these are expected to be global key ranges, but there are other functions like `Reader.MVCCIterate` where it isn't obvious (though the callee could look at both start and end and if they are not both local or both global, return an error).

What do you think?
___
*[pkg/kv/kvserver/store.go, line 2163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTS_4tA6StYlatgact:-MObeWZD2Arba38KpaZF:b-y7lsrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2163)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Non-mvcc meaning lock-table keys? Does this include keys with a 0 timestamp?
</blockquote></details>

Yes, keys that cannot be parsed as MVCCKeys. 0 timestamp keys are fine. I've rephrased the comment.
___
*[pkg/kv/kvserver/store.go, line 2165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTTGZC5UlEwKxWCMuk:-MObgj5407DySrMU-AAg:btoi87p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2165)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Yes, there can be more than 10. These keys mirror `system.settings`.
</blockquote></details>

Glad you asked about that suspicious logic -- I'd eyed it suspiciously and ignored it. I've added a comment here and a TODO(tschottdorf).
___
*[pkg/storage/intent_interleaving_iter.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTZHLd1Io3eTv623MN:-MObhoCU4KyVYuwNJfUe:b-il8a0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/intent_interleaving_iter.go#L146)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:thumbsup: cloning just means creating two iterators with the same sequence number, right?
</blockquote></details>

It will use the same memtables and sstables and sequence number. It is insufficient to use the same sequence number since an intervening compaction may have collapsed a.SET.10, a.SET.5 into a.SET.10, so having a sequence number of 7 will not help the clone see a.SET.5.
___
*[pkg/storage/mvcc_incremental_iterator_test.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVA0C0RJabvzQjL_O:-MObigRDCBJuMvmOdbDQ:b-t7p1ef) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/mvcc_incremental_iterator_test.go#L193)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we rename `keyMin` to `localMin` to avoid the possibility of confusion?
</blockquote></details>

Did you mean this local variable `keyMin`? It is the `globalMin`. Maybe I am misunderstanding the suggestion.
___
*[pkg/storage/mvcc_test.go, line 1037 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVvWc0QECtK9kTqh0:-MObjDRW5lC-OE-7HXjY:bjvrhtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e926b4b04da57171d0fb38159a602da3a2f19f/pkg/storage/mvcc_test.go#L1037)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why did this need to change? Something to do with the MVCCDelete and its interaction with the iterator?
</blockquote></details>

Correct. One cannot expect batch iterators to see mutations that were made after the iterator was created without doing a seek. I've now added a comment on the `Batch` interface in `engine.go`. This test was testing a stronger behavior than what we need. It broke because the `EngineIterator` in the `intentInterleavingIterator` that was iterating over the locks was already exhausted, so was not going to see the new intent.
___
*[pkg/storage/pebble_iterator.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTWIPK5lXVBs3eLuVB:-MOc02Gu5CgQvYjHLC5h:blc33c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/pebble_iterator.go#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Is there a benefit to the int vs. a bool? I ask because the int is going to increase the size of this struct while a bool would pack into the same word as `mvccDone`. In fact, you could rearrange this struct to keep all the bools next to each other if you wanted to try to keep the total size down. 
</blockquote></details>

Switched to bool and rearranged, which brings down the size from 240 to 224.
___
*[pkg/storage/pebble_iterator.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTXjJg46yPgW1JUgad:-MObuH708e-6_V8YeazI:bapbmx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/pebble_iterator.go#L246)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What is needed to fix the signatures? An error return value on `Key` and `UnsafeKey`? Or do you plan to pull key parsing into `Valid` for cases where it's needed?
</blockquote></details>

`Key` and `UnsafeKey` would return an error, just like `EngineIterator.EngineKey`. There is already a TODO to get rid of `Valid`, by returning `(bool, error)` from the iteration calls, so probably do both at the same time.
___
*[pkg/storage/testdata/intent_interleaving_iter/basic, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTYjv2EzpfipUqZzUo:-MObvO61E4cUGb0VfJJS:b-3l6mwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/testdata/intent_interleaving_iter/basic#L448)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this comment correct? Isn't this confirming that the _lock table iterator_ does not iterate _out of_ the lock table key space despite no upper bound?
</blockquote></details>

The MVCCKeys here are local keys, so the lower bound prevents iteration of the lock table iterator into these local keys.  


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-:-MOcOmuF521af6buGdml:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/consistency_queue_test.go, line 446 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MObc2K86qt6da1U-2qz:-MOcMzUJ5ygeduLxYgaj:bu55q0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/consistency_queue_test.go#L446)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I've added a TODO/question in `mvcc.go` where `ComputeStatsForRange` is declared. If we consider the first KV range to start at `KeyMin`, then this change in tests may not be sufficient and we may have an error-prone situation whenever we add new callers for any function that expects the range to be only global or only local.

For functions like `ComputeStatsForRange` it is viable to do this compensation at the callee since these are expected to be global key ranges, but there are other functions like `Reader.MVCCIterate` where it isn't obvious (though the callee could look at both start and end and if they are not both local or both global, return an error).

What do you think?
</blockquote></details>

So you're suggesting that mvcc operations that operate on key spans should assert that the spans are either entirely local or entirely global? That seems pretty reasonable to me, as it's never clear what semantics a caller would want out of a scan across this boundary.
___
*[pkg/storage/intent_interleaving_iter.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTZHLd1Io3eTv623MN:-MOcLTKt4uYwtjWbwglX:bx40ouy) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/intent_interleaving_iter.go#L146)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

It will use the same memtables and sstables and sequence number. It is insufficient to use the same sequence number since an intervening compaction may have collapsed a.SET.10, a.SET.5 into a.SET.10, so having a sequence number of 7 will not help the clone see a.SET.5.
</blockquote></details>

Got it, thanks for the explanation.
___
*[pkg/storage/mvcc.go, line 3432 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOcNBKS7KArajqCQlM4:-MOcNBKS7KArajqCQlM5:bbsectu) ([raw file](https://github.com/cockroachdb/cockroach/blob/525fffd88ff7d519b9faba50894550e39eb7f975/pkg/storage/mvcc.go#L3432)):*
> ```Go
> // the total age of all intents.
> //
> // TODO(sumeer): the comment about "adjusted to avoid counting local keys" is
> ```

Yeah, I'm not quite sure what this is referencing. This comment was added in https://github.com/cockroachdb/cockroach/commit/768989ba4274d8e02d455aeff38de620619d36c9#diff-7fbe8fc69d8e2074417269cb9e77af42bfbe2bdec507593b2b8760bf9f447ab4R2339, but even there, we didn't do any "adjusting" of the start key in either the Go or C++ implementations of this function. I think this might be referring to https://github.com/cockroachdb/cockroach/pull/3155/files#diff-d784010ce8fff8d73c814cae431a34c64f3fff2c3a8085d4154afb83e4ac2911R49, which is logic that still lives to this day.
___
*[pkg/storage/mvcc_incremental_iterator_test.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVA0C0RJabvzQjL_O:-MOcOp3LCZsGkpgNuXQb:b-9vpjio) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/mvcc_incremental_iterator_test.go#L193)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Did you mean this local variable `keyMin`? It is the `globalMin`. Maybe I am misunderstanding the suggestion.
</blockquote></details>

Oh sorry, I meant `localMax`.
___
*[pkg/storage/mvcc_test.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVlYzDQeyYyI_xs-H:-MOcOuzP8z4roRWSPJsW:bdpu218) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/mvcc_test.go#L48)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same question here about `localMin`.
</blockquote></details>

Same adjustment.
___
*[pkg/storage/pebble_iterator.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTWIPK5lXVBs3eLuVB:-MOcLfbG2ubk6NGN8rtz:b-n8y1hi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/pebble_iterator.go#L59)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Switched to bool and rearranged, which brings down the size from 240 to 224.
</blockquote></details>

:thumbsup: If you like the int better, you could also use an `int8`.
___
*[pkg/storage/testdata/intent_interleaving_iter/basic, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTYjv2EzpfipUqZzUo:-MOcM0Ys04h5eEB3I2GX:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/testdata/intent_interleaving_iter/basic#L448)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The MVCCKeys here are local keys, so the lower bound prevents iteration of the lock table iterator into these local keys.  
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57507)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/store.go, line 2165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTTGZC5UlEwKxWCMuk:-MOf1PCu6Q0PDvgNqsY3:bddqu5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2165)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Glad you asked about that suspicious logic -- I'd eyed it suspiciously and ignored it. I've added a comment here and a TODO(tschottdorf).
</blockquote></details>

Oh, only saw now that this was in the old code as well! Goot catch. Please remove the `maxKeys` restriction, it's bogus (premature optimization from ca 2014). Btw it's `tbg` these days but the TODO will be going away as well. :)


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57507)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/consistency_queue_test.go, line 446 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MObc2K86qt6da1U-2qz:-MOi-L4b48J2EA4ylNV8:bvlvusp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/consistency_queue_test.go#L446)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

So you're suggesting that mvcc operations that operate on key spans should assert that the spans are either entirely local or entirely global? That seems pretty reasonable to me, as it's never clear what semantics a caller would want out of a scan across this boundary.
</blockquote></details>

We also do such computations over non-Pebble iterators like an `sstIterator` where it is reasonable for someone to shove both local and global keys for a range into a single sstable (from a correctness perspective). So I didn't add any error checking here and instead added a `panic` to `newIntentInterleavingIter`.
___
*[pkg/kv/kvserver/store.go, line 2165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTTGZC5UlEwKxWCMuk:-MOhrRA0-dOfytwU7n2I:b-yslgxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2165)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Oh, only saw now that this was in the old code as well! Goot catch. Please remove the `maxKeys` restriction, it's bogus (premature optimization from ca 2014). Btw it's `tbg` these days but the TODO will be going away as well. :)
</blockquote></details>

Done. I rearranged the code to make this work. Is there an existing test that directly or indirectly exercises this?
___
*[pkg/storage/mvcc.go, line 3432 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOcNBKS7KArajqCQlM4:-MOhuEsT3PfaaqpNx-5U:b6rjsva) ([raw file](https://github.com/cockroachdb/cockroach/blob/525fffd88ff7d519b9faba50894550e39eb7f975/pkg/storage/mvcc.go#L3432)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah, I'm not quite sure what this is referencing. This comment was added in https://github.com/cockroachdb/cockroach/commit/768989ba4274d8e02d455aeff38de620619d36c9#diff-7fbe8fc69d8e2074417269cb9e77af42bfbe2bdec507593b2b8760bf9f447ab4R2339, but even there, we didn't do any "adjusting" of the start key in either the Go or C++ implementations of this function. I think this might be referring to https://github.com/cockroachdb/cockroach/pull/3155/files#diff-d784010ce8fff8d73c814cae431a34c64f3fff2c3a8085d4154afb83e4ac2911R49, which is logic that still lives to this day.
</blockquote></details>

I've deleted the stale comment. 
___
*[pkg/storage/mvcc_incremental_iterator_test.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVA0C0RJabvzQjL_O:-MOi04aBAjhxLs8OKqTG:bmpbzbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/mvcc_incremental_iterator_test.go#L193)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Oh sorry, I meant `localMax`.
</blockquote></details>

Ah yes, that makes sense. Done.
___
*[pkg/storage/mvcc_test.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTVlYzDQeyYyI_xs-H:-MOi0KpwE47M50NI1O7-:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/mvcc_test.go#L48)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same adjustment.
</blockquote></details>

Done
___
*[pkg/storage/pebble_iterator.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTWIPK5lXVBs3eLuVB:-MOi0MRWBp5gV0-_gaAo:b-i16a6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/storage/pebble_iterator.go#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:thumbsup: If you like the int better, you could also use an `int8`.
</blockquote></details>

Left it as bool. Slightly prefer an uninitialized variable to have a valid value, given that Go does not force callers to use a "constructor".


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57507)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/store.go, line 2165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTTGZC5UlEwKxWCMuk:-MOjy9yjCqMI7WMn08pe:bpuyicc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2165)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Done. I rearranged the code to make this work. Is there an existing test that directly or indirectly exercises this?
</blockquote></details>

I doubt it. This behavior wasn't ever useful, btw. We only call this method after having established that there isn't a StoreIdent here, so it's really just a sanity check that may not even be worth it.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-:-MOlL2wV2m3PhJN3eBIP:bodi8iv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/store.go, line 2165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57507#-MOTTGZC5UlEwKxWCMuk:-MOlKM-V4NS2tOkkBDYU:b2nhb2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cc7473ac4ed12c84cdef2840cf15d8552dc7837/pkg/kv/kvserver/store.go#L2165)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I doubt it. This behavior wasn't ever useful, btw. We only call this method after having established that there isn't a StoreIdent here, so it's really just a sanity check that may not even be worth it.
</blockquote></details>

Ack


<!-- Sent from Reviewable.io -->

sumeerbhola commented:
bors r+
